File: src/fetchers/aviasales.py
"""
Aviasales / Travelpayouts API
Registro: https://travelpayouts.com/
Documentação: https://support.travelpayouts.com/hc/en-us/articles/203956085

Endpoint de calendário: retorna o preço mais barato por dia do mês inteiro em 1 chamada.
Ideal para varredura eficiente dos próximos 12 meses.
"""
import os
import httpx
from datetime import date, datetime
from dotenv import load_dotenv

load_dotenv()

AVIASALES_TOKEN = os.getenv("AVIASALES_TOKEN", "")
BASE_URL = "https://api.travelpayouts.com"


def _is_configured() -> bool:
    return bool(AVIASALES_TOKEN)


def fetch_month_calendar(origin: str, destination: str, year: int, month: int) -> list[dict]:
    """
    Retorna o preço mais barato para cada dia do mês (só ida).
    1 chamada de API retorna preços de um mês inteiro — muito eficiente.
    """
    if not _is_configured():
        return []

    depart_date = f"{year}-{month:02d}"

    try:
        with httpx.Client(timeout=30) as client:
            response = client.get(
                f"{BASE_URL}/v1/prices/calendar",
                params={
                    "origin": origin,
                    "destination": destination,
                    "depart_date": depart_date,
                    "currency": "brl",
                    "show_to_affiliates": "false",
                    "token": AVIASALES_TOKEN,
                },
            )
            response.raise_for_status()
            data = response.json()
    except Exception as e:
        print(f"[Aviasales] Erro calendário {origin}→{destination} {depart_date}: {e}")
        return []

    if not data.get("success") or not data.get("data"):
        print(f"[Aviasales] Sem dados para {origin}→{destination} {depart_date}: success={data.get('success')}, data={data.get('data')}, raw={data}")
        return []

    results = []
    for date_str, info in data["data"].items():
        try:
            flight_date = datetime.strptime(date_str, "%Y-%m-%d").date()
            price_brl = float(info.get("price", 0))
            if price_brl <= 0:
                continue

            results.append({
                "flight_date": flight_date,
                "return_date": None,
                "airline": info.get("airline", "?"),
                "price_brl": price_brl,
                "price_eur": None,  # Aviasales não retorna EUR neste endpoint
                "stops": info.get("transfers", 0),
                "duration_minutes": None,
                "source": "aviasales",
            })
        except Exception:
            continue

    return results


def fetch_cheap_prices(origin: str, destination: str) -> list[dict]:
    """
    Retorna os voos mais baratos disponíveis (sem data específica).
    Útil para ter uma visão geral rápida de oportunidades.
    """
    if not _is_configured():
        return []

    try:
        with httpx.Client(timeout=30) as client:
            response = client.get(
                f"{BASE_URL}/v1/prices/cheap",
                params={
                    "origin": origin,
                    "destination": destination,
                    "currency": "brl",
                    "token": AVIASALES_TOKEN,
                },
            )
            response.raise_for_status()
            data = response.json()
    except Exception as e:
        print(f"[Aviasales] Erro cheap prices {origin}→{destination}: {e}")
        return []

    if not data.get("success") or not data.get("data"):
        return []

    results = []
    dest_data = data["data"].get(destination, {})
    for trip_class, info in dest_data.items():
        try:
            dep_date_str = info.get("depart_date", "")
            if not dep_date_str:
                continue
            flight_date = datetime.strptime(dep_date_str, "%Y-%m-%d").date()
            price_brl = float(info.get("price", 0))
            if price_brl <= 0:
                continue

            results.append({
                "flight_date": flight_date,
                "return_date": None,
                "airline": info.get("airline", "?"),
                "price_brl": price_brl,
                "price_eur": None,
                "stops": info.get("transfers", 0),
                "duration_minutes": None,
                "source": "aviasales",
            })
        except Exception:
            continue

    return results

File: backend/src/collectors/aviasales.ts
import axios from 'axios'
import { FlightRecord } from './types'

const BASE_URL = 'https://api.travelpayouts.com/v1/prices'

function isConfigured(): boolean {
  return !!process.env.AVIASALES_TOKEN
}

interface CalendarDay {
  price: number
  airline: string
  transfers: number
}

interface CalendarResponse {
  success: boolean
  data: Record<string, CalendarDay>
}

export interface FetchMonthCalendarResult {
  records: FlightRecord[]
  error?: string
}

function briefErr(err: unknown): string {
  if (err instanceof Error) return err.message
  return String(err)
}

export async function fetchMonthCalendar(
  origin: string,
  destination: string,
  yearMonth: string // YYYY-MM
): Promise<FetchMonthCalendarResult> {
  if (!isConfigured()) return { records: [] }

  try {
    const response = await axios.get<CalendarResponse>(`${BASE_URL}/calendar`, {
      params: {
        origin,
        destination,
        depart_date: yearMonth,
        currency: 'brl',
        show_to_affiliates: false,
        token: process.env.AVIASALES_TOKEN,
      },
      timeout: 15000,
    })

    if (!response.data.success || !response.data.data) return { records: [] }

    const records = Object.entries(response.data.data)
      .map(([dateStr, day]) => ({
        flightDate: new Date(dateStr + 'T00:00:00'),
        returnDate: null,
        airline: day.airline || null,
        priceBrl: day.price,
        priceEur: null,
        stops: day.transfers ?? 0,
        durationMinutes: null,
        source: 'aviasales' as const,
      }))
      .filter((r) => r.priceBrl > 0)
    return { records }
  } catch (err) {
    const msg = briefErr(err)
    console.error(`[Aviasales] Erro ao buscar ${origin}→${destination} ${yearMonth}: ${msg}`)
    return { records: [], error: msg }
  }
}

interface CheapResponse {
  success: boolean
  data: Record<string, Record<string, { depart_date: string; price: number; airline: string; transfers: number }>>
}

export async function fetchCheapPrices(
  origin: string,
  destination: string
): Promise<FlightRecord[]> {
  if (!isConfigured()) return []

  try {
    const response = await axios.get<CheapResponse>(`${BASE_URL}/cheap`, {
      params: {
        origin,
        destination,
        currency: 'brl',
        token: process.env.AVIASALES_TOKEN,
      },
      timeout: 15000,
    })

    if (!response.data.success || !response.data.data) return []

    const destData = response.data.data[destination]
    if (!destData) return []

    return Object.values(destData)
      .map((entry) => ({
        flightDate: new Date(entry.depart_date + 'T00:00:00'),
        returnDate: null,
        airline: entry.airline || null,
        priceBrl: entry.price,
        priceEur: null,
        stops: entry.transfers ?? 0,
        durationMinutes: null,
        source: 'aviasales' as const,
      }))
      .filter((r) => r.priceBrl > 0)
  } catch (err) {
    console.error(`[Aviasales] Erro cheap ${origin}→${destination}:`, err)
    return []
  }
}
